datalad/interface/utils.py

```diff
@@ -761,3 +765,105 @@ def filter_unmodified(content_by_ds, refds, since):
                         # or a modified path under a given directory
                         or any(m.startswith(_with_sep(p)) for p in paths_refds))]
     return keep
+
+
+def eval_results(func):
+    """Decorator for return value evaluation of datalad commands.
+
+    Note, this decorator is only compatible with commands that return
+    status dict sequences!
+
+    Two basic modes of operation are supported: 1) "generator mode" that
+    `yields` individual results, and 2) "list mode" that returns a sequence of
+    results. The behavior can be selected via the
+    `datalad.api.return-generator` configuration variable. Default is "list
+    mode".
+
+    This decorator implements common functionality for result rendering/output,
+    error detection/handling, and logging (TODO).
+
+    Result rendering/output can be triggered via the
+    `datalad.api.result-render-mode` configuration variable. Supported modes
+    are: 'json' (one object/dict per line, like git-annex), 'simple' (TODO,
+    tailored output formating provided by each command class, if any).
+
+    Error detection works by inspecting the `status` item of all result
+    dictionaries. Any occurrence of a status other than 'ok' or 'notneeded' will
+    cause an exception to be raised. TODO: The type of exception depends on the
+    nature of complete set of errors found. If all errors have a single unique
+    type, the final exception will be of this type too. In case of a heterogeneous
+    set of error a compound exception will be raised that includes information
+    on all individual errors that occurred.
+
+    Status messages will be logged automatically, by default the following
+    association of result status and log channel will be used: 'ok' (debug),
+    'notneeded' (debug), 'impossible' (warning), 'error' (error).  Logger
+    instances included in the results are used to capture the origin of a
+    status report.
+    """
+    want_generator = dlcfg.getbool('datalad.api', 'return-generator', False)
```

Comment:
mental note to myself possibly
- either all decorated commands return a generator, or just return a list. all the time
- we will need travis setup to test correct internal operation in both modes
- since decision is made once for the entire process, usefulness is probably only for command line invocations of separate datalad commands 
- I do not think there is an easy way (if any?) to make it 'run time' configurable per a command invocation by adding some kwarg to the function which decorator would consume and either return a generator or the list of results... actually, why not?
```python
$> python /tmp/f.py
<generator object gen_func at 0x7f7d584f08c0>
[1, 2]

$> cat /tmp/f.py

def gen_func():
   yield 1
   yield 2

def dynamic(f, what):
   res = f()
   return res if what == 'generator' else list(res)

print dynamic(gen_func, 'generator')
print dynamic(gen_func, 'list')

```
so probably it could be made configurable at run time by decoration checking some kwarg which could augment default behavior... or I am missing smth